packages/patterns-react/src/patterns/TableOfContents/TableOfContents.js

```diff
@@ -0,0 +1,271 @@
+/**
+ * Copyright IBM Corp. 2016, 2018
+ *
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import PropTypes from 'prop-types';
+import React from 'react';
+import { settings as ddsSettings } from '@carbon/ibmdotcom-utilities';
+import { settings } from 'carbon-components';
+import classNames from 'classnames';
+import { featureFlag } from '@carbon/ibmdotcom-utilities';
+import { TABLE_OF_CONTENTS } from '../../internal/FeatureFlags';
+import { TableOfContents20 } from '@carbon/icons-react';
+import { Layout } from '@carbon/ibmdotcom-react';
+
+const { stablePrefix } = ddsSettings;
+const { prefix } = settings;
+
+/**
+ * Table of Contents pattern
+ *
+ * @param {object} props props object
+ * @param {string} props.layoutType layout type property ( 1/3 - 1/4 )
+ * @param {object} props.menuItems menu items object which includes title and id
+ * @param {element} props.children children property of table of contents pattern
+ * @param {object} props.marginTop margin top property which adds margin top value to layout
+ * @param {object} props.marginBottom margin bottom property which adds margin top value to layout
+ * @returns {*} Table of Content pattern
+ */
+class TableOfContents extends React.Component {
+  menuRef = React.createRef();
+  state = {
+    selectedItem: 1,
+    selectedTitle: '',
+  };
+
+  /**
+   *
+   *
+   * @memberof TableOfContents
+   */
+  componentDidMount() {
+    window.addEventListener('scroll', this.isInViewport);
+  }
+
+  /**
+   *
+   *
+   * @memberof TableOfContents
+   */
+  componentWillUnmount() {
+    window.removeEventListener('scroll', this.isInViewport);
+  }
+
+  /**
+   * Check whether content is in view
+   *
+   * @memberof TableOfContents
+   */
+  isInViewport = () => {
+    const items = document.querySelectorAll('[data-driverlocation]');
```

Comment:
use `window-or-root` when referencing `document`

Reply:
The name `data-driverlocation` seems odd. Should it be something like `data-toc-anchor`?

I'm wondering also if this should be tightly coupled like this. The `menuItems` prop is passing in the list of menu items with corresponding `id`'s . I'm wondering if we could have them as anchor names instead? The content on the right would simple have `<a name="[anchorID]">`, and for menuItems it would be the `title` and `anchorID`. The AnchorID would also double as a standard HTML jump link.

Reply:
I see what you mean but it is the attribute name that DDP team wants to use for link reference.

Reply:
@sannminwin that makes this pattern specific to Drupal then, which is something we should avoid.

Reply:
> The name `data-driverlocation` seems odd. Should it be something like `data-toc-anchor`?
> 
> I'm wondering also if this should be tightly coupled like this. The `menuItems` prop is passing in the list of menu items with corresponding `id`'s . I'm wondering if we could have them as anchor names instead? The content on the right would simple have `<a name="[anchorID]">`, and for menuItems it would be the `title` and `anchorID`. The AnchorID would also double as a standard HTML jump link.

I see,
content on the right side is unknow to us. The only identifier is data-driverlocation attribute and this generated content won't have id or being able to add it. 
one thing we can do to loosely couple the selector is passing in selector property into pattern. Author will be able to choose appropiate selector and pass in as prop.

Reply:
and we can use anchorID  as prop name for menuItems.

Reply:
> content on the right side is unknow to us. The only identifier is data-driverlocation attribute and this generated content won't have id or being able to add it.
> one thing we can do to loosely couple the selector is passing in selector property into pattern. Author will be able to choose appropiate selector and pass in as prop.

I think passing in the selector info as part of the props might work, then that means the left nav doesn't have to worry about what to detect. So the `menuItems` would then pass in the `title` and the `uniqueSelector` for each item instead of `id`. The other thought, it could be anchor name. The left side would then be content with things like `<a name="MyAnchorName"></a><h2>Lorem Ipsum</h2>`, then the `menuItems` would then pass in as:

```javascript
[
  {
    anchorName: 'MyAnchorName',
    title: 'Lorem Ipsum',
  },
  ...
]
```

And query selectors would be `root.document.querySelector("a[name='${menuItems[i].anchorName}']")`

Reply:
I am coming to your desk to talk about it.

Reply:
As discussed, we'd like to use anchor links with the name attribute (for pure HTML jump links) as a way to identify the actual jump link in the content section. The TOC component can detect based on the anchor name that's passed in as the menuItems. We could even explore if the TOC component can just detect all of the anchor names in the content well so the TOC links do not have to be manually defined, but that can be handled separately (but can lead to that).

![MVIMG_20191113_170710](https://user-images.githubusercontent.com/1641214/68808703-4a238b00-0638-11ea-83c5-40815136c97b.jpg)